Allow asymmetric in/out buffer lengths to conserve RAM

projectgus opened the pull request:
For the situation where the mbedTLS device has limited RAM, but the
other end of the connection doesn't support the max_fragment_length
extension. To be spec-compliant, mbedTLS has to keep a 16384 byte
incoming buffer. However the outgoing buffer can be made smaller without
breaking spec compliance, and we save some RAM.

If buffer lengths are kept equal, this should make zero changes to the compiled library.

See comments in include/mbedtls/config.h for some more details.

(The lower limit of outgoing buffer size is the buffer size used during
handshake/cert negotiation. As the handshake is half-duplex it might
even be possible to store this data in the "incoming" buffer during the
handshake, which would save even more RAM - but it would also be a lot
hackier and error-prone. I didn't really explore this possibility, but
thought I'd mention it here in case someone sees this later on a mission
to jam mbedTLS into an even tinier RAM footprint.)

projectgus commented:
(Testing this proved a bit tricky because most internet HTTPS servers don't seem to send full 16KB messages, even for static content. Although they sometimes might send them, I'm guessing fragmentation between the TLS gateway and the source the message size smaller (I saw 8-10KB mostly.) 

The command I ended up using for local testing was
`dd if=/dev/zero bs=32K | openssl s_server -key key.pem -cert cert.pem -tls1_2 -accept 2000`

projectgus commented:
This may seem like a micro-optimisation depending on your perspective. On my target platform, with 80KB of RAM, clawing back 12KB  (using a 4KB OUT buffer) while keeping spec compliance (16KB IN buffer) was a big win.

I understand that the complexity/savings tradeoff may be different for use cases with more RAM available, but I thought I'd send the PR for consideration anyhow.

projectgus commented:
I'm not able to reproduce this doxygen failure that's happening on Travis:
```
$ tests/scripts/doxygen.sh

/home/travis/build/ARMmbed/mbedtls/include/mbedtls/config.h:2545: warning: documentation for unknown define MBEDTLS_SSL_MAX_CONTENT_LEN found.

/home/travis/build/ARMmbed/mbedtls/include/mbedtls/config.h:2551: warning: documentation for unknown define MBEDTLS_SSL_IN_CONTENT_LEN found.

/home/travis/build/ARMmbed/mbedtls/include/mbedtls/config.h:2557: warning: documentation for unknown define MBEDTLS_SSL_OUT_CONTENT_LEN found.

FAIL
```

If I run it locally it succeeds. It seems like the config generation script should find these commented constants and enable them. Any clues?

pjbakker commented:
Might depend on the version of doxygen?

projectgus commented:
@pjbakker - maybe? The script which comments macros is Perl though - scripts/config.pl - and surely it wouldn't be Perl version dependent. Hmm.

frozeus commented:
Hi,

cherry-pick this commit on the development branch and i don't have these warnings.

```
$ doxygen --version
1.8.11
```

projectgus commented:
Thanks @frozeus , good to know it's not just me where it passes. I rebased against development just now, but same doxygen failure. :(.

I will try to make time to spin up an Ubuntu 14.04 VM and reproduce the steps Travis runs with the same versions of everything.

frozeus commented:
The version on the server is :
`Get:1 http://us.archive.ubuntu.com/ubuntu/ precise/main doxygen amd64 1.7.6.1-2ubuntu1 [2,427 kB]`

You can get it here : 
https://launchpad.net/ubuntu/+source/doxygen/1.7.6.1-2ubuntu1/+build/3392241

With this version i have the same output :
```
./tests/scripts/doxygen.sh 
/home/clement/Documents/mbedtls/include/mbedtls/config.h:2551: warning: documentation for unknown define MBEDTLS_SSL_MAX_CONTENT_LEN found.
/home/clement/Documents/mbedtls/include/mbedtls/config.h:2557: warning: documentation for unknown define MBEDTLS_SSL_IN_CONTENT_LEN found.
/home/clement/Documents/mbedtls/include/mbedtls/config.h:2563: warning: documentation for unknown define MBEDTLS_SSL_OUT_CONTENT_LEN found.
FAIL
```

projectgus commented:
Finally got back to this, sorry for the very extended delay!

Fixed the doxygen syntax, rebased on latest master.

projectgus commented:
Rebased on latest master again.

projectgus commented:
(Sorry, I missed that when I rebased in October I accidentally created an unused variable. Rebased again.)

gilles-peskine-arm commented on include/mbedtls/compat-1.3.h:
```diff
@@ -1377,7 +1377,6 @@
 #define SSL_ANTI_REPLAY_ENABLED MBEDTLS_SSL_ANTI_REPLAY_ENABLED
 #define SSL_ARC4_DISABLED MBEDTLS_SSL_ARC4_DISABLED
 #define SSL_ARC4_ENABLED MBEDTLS_SSL_ARC4_ENABLED
-#define SSL_BUFFER_LEN MBEDTLS_SSL_BUFFER_LEN
```
Unfortunately, `SSL_BUFFER_LEN` was a (semi-)public API PolarSSL in 1.3, so we can't remove it from the compatibility translation. It needs to be implemented in a way that works when neither `MBEDTLS_SSL_IN_CONTENT_LEN` nor `MBEDTLS_SSL_OUT_CONTENT_LEN` are defined (since these did not exist in 1.3).

projectgus replied:
Have implemented as the minimum of `MBEDTL_SSL_IN_BUFFER_LEN` & `MBEDTLS_OUT_BUFFER_LEN` (which probably have equal values in any 1.3-compatible setup, but I figure it can't hurt to be certain in case of some mixed codebase.

gilles-peskine-arm commented:
This is a useful change which improves the RAM consumption significantly on small devices when the input and output buffers are not shared. I am satisfied that the change is desirable.

I am happy with the design which preserves backward compatibility except in internal APIs, with one easily-fixed exception regarding upgrading from 1.3 (https://github.com/ARMmbed/mbedtls/pull/536#discussion_r177468620).

This needs some testing to ensure that we use the correct `MBEDTLS_SSL_xxx_CONTENT_LEN` in each place. This means building the library with memsan and with different in/out buffer sizes and running some intelligently-chosen test cases in `ssl-opt.sh` (perhaps existing test cases, perhaps new ones).

DShabroff commented:
This changes are workig well. But there is 5 extra points in ssl_tls.c need to change in my case. This are points:

`//1330-----------------------------------------------
if(ssl->out_msglen > MBEDTLS_SSL_OUT_CONTENT_LEN) //MBEDTLS_SSL_MAX_CONTENT_LEN -> MBEDTLS_SSL_OUT_CONTENT_LEN
{
MBEDTLS_SSL_DEBUG_MSG(1, ("Record content %u too large, maximum %d",
(unsigned) ssl->out_msglen,
MBEDTLS_SSL_OUT_CONTENT_LEN));//MBEDTLS_SSL_MAX_CONTENT_LEN -> MBEDTLS_SSL_OUT_CONTENT_LEN
return(MBEDTLS_ERR_SSL_BAD_INPUT_DATA);
}
//2807-----------------------------------------------
/* Make room for the additional DTLS fields */
if(MBEDTLS_SSL_OUT_CONTENT_LEN - ssl->out_msglen < 8) //MBEDTLS_SSL_MAX_CONTENT_LEN -> MBEDTLS_SSL_OUT_CONTENT_LEN
{
MBEDTLS_SSL_DEBUG_MSG(1, ("DTLS handshake message too large: "
"size %u, maximum %u",
(unsigned) (ssl->in_hslen - 4),
(unsigned) (MBEDTLS_SSL_IN_CONTENT_LEN - 12))); //MBEDTLS_SSL_MAX_CONTENT_LEN -> MBEDTLS_SSL_IN_CONTENT_LEN
return(MBEDTLS_ERR_SSL_BAD_INPUT_DATA);
}
//7196-----------------------------------------------
size_t max_len = MBEDTLS_SSL_OUT_CONTENT_LEN;//MBEDTLS_SSL_MAX_CONTENT_LEN -> MBEDTLS_SSL_OUT_CONTENT_LEN
//-----------------------------------------------`

itayzafrir commented:
Hi @projectgus, please take a look at https://github.com/ARMmbed/mbedtls/pull/536/commits/7ff5b08a6dc1f8304fc5f58b00d944b346d8ab43#diff-995204bbb15401337db9b64bc314e1f6R2185
and also https://github.com/ARMmbed/mbedtls/pull/536/commits/7ff5b08a6dc1f8304fc5f58b00d944b346d8ab43#diff-995204bbb15401337db9b64bc314e1f6R2194

I think that instead of MBEDTLS_SSL_IN_BUFFER_LEN it should be MBEDTLS_SSL_IN_CONTENT_LEN ?

projectgus commented:
@itayzafrir Oops, thanks for that. Fixed!

Thanks also for reminding me about @gilles-peskine-arm request for "intelligently chosen test cases". Will rework so only the large buffer & fragment tests are run with asymmetric buffers, rather than all compat.sh & ssl-opt.sh tests.

projectgus commented:
Hi @hanno-arm @gilles-peskine-arm @DShabroff @itayzafrir ,

Updated, comments addressed and custom tests added for different buffer sizes (both reduced overall buffer sizes, and asymmetric buffers).

Also addressed @sbutcher-arm's comments as made on the copy of this patch that was submitted #1568 

Please take a look. :)

itayzafrir commented on tests/ssl-opt.sh:
```diff
@@ -1474,18 +1487,18 @@ run_test    "Max fragment length, DTLS: enabled, default, larger message" \
+            "$P_CLI debug_level=3 request_size=$(expr $MAX_CONTENT_LEN + 1)" \
             0 \
             -C "Maximum fragment length is 16384" \
             -S "Maximum fragment length is 16384" \
```
Shouldn't it be:
`-C "Maximum fragment length is $MAX_CONTENT_LEN" \`
`-S "Maximum fragment length is $MAX_CONTENT_LEN" \`
?

projectgus replied:
No, because in this test (and the one mentioned below) MBEDTLS_SSL_MAX_FRAGMENT_LENGTH is disabled on the mbedTLS side (see a few lines above), so mbedTLS can't negotiate a smaller fragment length than the default (even if the other end requests it).

I'll add a comment to this effect, as it's not immediately obvious what's happening here.

itayzafrir commented on tests/ssl-opt.sh:
```diff
@@ -1474,18 +1487,18 @@ run_test    "Max fragment length, DTLS: enabled, default, larger message" \
+            "$P_CLI debug_level=3 dtls=1 request_size=$(expr $MAX_CONTENT_LEN + 1)" \
             1 \
             -C "Maximum fragment length is 16384" \
             -S "Maximum fragment length is 16384" \
```
Shouldn't it be:
`-C "Maximum fragment length is $MAX_CONTENT_LEN" \`
`-S "Maximum fragment length is $MAX_CONTENT_LEN" \`
?

projectgus commented on include/mbedtls/ssl_internal.h:
```diff
@@ -158,17 +173,52 @@
+ */
+#define MBEDTLS_SSL_BUFFER_LEN  MBEDTLS_SSL_IN_BUFFER_LEN
+#define MBEDTLS_SSL_PAYLOAD_LEN MBEDTLS_SSL_IN_PAYLOAD_LEN
+#endif
```
Good point, will drop these - safer.

projectgus commented:
Thanks @itayzafrir for the comments, I should have looked at your version of the PR more closely! Have updated in accordance, PTAL.

> Regarding the Memsan build, I'm not sure it's completely necessary. [...]

You're totally right and I totally agree. Have removed the MSan version of the tests.

> There might be an issue with file compat-1.3.h since MBEDTLS_SSL_BUFFER_LEN isn't defined anymore. 

Yes... I'm not sure whether it's safe to assume ssl_internal.h will be included before compat-1.3.h, rather than after. So for this reason I've implemented it as a ternary expression, [see comment here](https://github.com/ARMmbed/mbedtls/pull/536/commits/8808bad06bf3bec7bc5d6a4986e6a877bbdfae0e#r177468620). Let me know what you think.

itayzafrir commented:
Hey @projectgus, I've reviewed your latest commits, looking good to me.

projectgus commented:
Rebased again to resolve conflicts.

@k-stachowiak @hanno-arm @Patater @sbutcher-arm This PR will celebrate its second birthday next month. It would be great to know before then if it's a viable candidate for merging. :)

sbutcher-arm commented on include/mbedtls/config.h:
```diff
@@ -2798,7 +2798,30 @@
 
 /* SSL options */
-//#define MBEDTLS_SSL_MAX_CONTENT_LEN             16384 /**< Maxium fragment length in bytes, determines the size of each of the two internal I/O buffers */
+//#define MBEDTLS_SSL_MAX_CONTENT_LEN             16384 /**< Maximum fragment length in bytes, determines the size of each of the two internal I/O buffers */
```
Style: these lines would be better wrapped at 80 columns. We break this rule elsewhere in the source code, but try not to.

projectgus replied:
I think the reason this is done for many items in config.h is that splitting a line which was commented with `//` is awkward, ie

```
//#define MBEDTLS_SSL_MAX_CONTENT_LEN             16384 /**< Maximum fragment length in bytes, determines the size of each
                                                             of the two internal I/O buffers */
```
... is a compiler error, and if you add a `//` on the second line then it breaks when you uncomment the macro if you don't also uncomment the second line.

Suggest that if you want to break this pattern then it'd be best to do it in a single change which reformats all config.h items to get the desired layout throughout (there a lot like this one, or with even longer inline comments after them).

sbutcher-arm commented on include/mbedtls/config.h:
```diff
@@ -2798,7 +2798,30 @@
+
+/** \def MBEDTLS_SSL_OUT_CONTENT_LEN
+ *
+ * Optionally override MBEDTLS_SSL_MAX_CONTENT_LEN for the outward TLS buffer only.
```
Documentation: This is a fragment, and it needs a verb! Suggestion:
`This option can override MBEDTLS_SSL_MAX_CONTENT_LEN for the outward TLS buffer only.`

gilles-peskine-arm replied:
“override” is a verb. If you change to add the subject “this option” then don't use “can” since this option _does_ override `MBEDTLS_SSL_MAX_CONTENT_LEN`.

gilles-peskine-arm replied:
But it isn't clear what “optionally” means here. My guess is that this means “if the line containing the `#define` directive is uncommented”, but I shouldn't have to guess. Elsewhere in the file we write “Uncomment to …”, please follow this style if you need to express that something isn't active by default. For the header of the description, though, just say what the feature does, and don't require the reader to know about another feature: “The size of the outward TLS buffer”.

sbutcher-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -143,12 +143,27 @@
-                        + MBEDTLS_SSL_MAC_ADD                    \
-                        + MBEDTLS_SSL_PADDING_ADD                \
-                        )
+#define MBEDTLS_SSL_PAYLOAD_OVERHEAD ( MBEDTLS_SSL_COMPRESSION_ADD       \
```
Style: Spacing is slightly out.

sbutcher-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -158,17 +173,43 @@
-#if MBEDTLS_SSL_PAYLOAD_LEN > 16384 + 2048
-#error Bad configuration - protected record payload too large.
+#if MBEDTLS_SSL_IN_CONTENT_LEN > MBEDTLS_SSL_MAX_CONTENT_LEN
+#error Bad configuration - incoming record content should not be larger than MBEDTLS_SSL_MAX_CONTENT_LEN.
```
I know this is replacing like for like, but all other `#error` directives encapsulate their error messages in strings. Offhand I'm not sure if there's any technical difference.

For consistency, it would be nice for it to be the same.

gilles-peskine-arm replied:
What follows `#error` must be lexically valid, so you can't portably have something like `#error You can't do this`. But an invalid `#error` directive is likely to terminate the compilation with a message containing what follows the `#error` directive, so it doesn't matter much.

sbutcher-arm commented on library/ssl_tls.c:
```diff
@@ -2173,8 +2183,7 @@ static int ssl_decompress_buf( mbedtls_ssl_context *ssl )
     ssl->transform_in->ctx_inflate.next_out = msg_post;
-    ssl->transform_in->ctx_inflate.avail_out = MBEDTLS_SSL_BUFFER_LEN -
-                                               header_bytes;
+    ssl->transform_in->ctx_inflate.avail_out = MBEDTLS_SSL_IN_BUFFER_LEN - header_bytes
```
Style: Format for 80 columns.

sbutcher-arm commented:
@gilles-peskine-arm / @Patater / @hanno-arm - I need approval from one of you, and then we're good to go!

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2125,7 +2133,8 @@ void mbedtls_ssl_conf_cert_req_ca_list( mbedtls_ssl_config *conf,
 /**
  * \brief          Set the maximum fragment length to emit and/or negotiate
- *                 (Default: MBEDTLS_SSL_MAX_CONTENT_LEN, usually 2^14 bytes)
+ *                 (Default: the smaller of MBEDTLS_SSL_IN_CONTENT_LEN &
```
No gratuitous and confusing (is it bitwise and?) abbreviations please: `&` → `and` (here and elsewhere)

gilles-peskine-arm commented on include/mbedtls/ssl_internal.h:
```diff
@@ -158,17 +173,43 @@
-#if MBEDTLS_SSL_PAYLOAD_LEN > 16384 + 2048
-#error Bad configuration - protected record payload too large.
+#if MBEDTLS_SSL_IN_CONTENT_LEN > MBEDTLS_SSL_MAX_CONTENT_LEN
+#error Bad configuration - incoming record content should not be larger than MBEDTLS_SSL_MAX_CONTENT_LEN.
```
Not a blocker since it's preexisting, but since `check_config.h` exists, I think all configuration checks should be in `check_config.h`.

gilles-peskine-arm commented on library/ssl_tls.c:
```diff
@@ -141,14 +141,24 @@ static void ssl_reset_retransmit_timeout( mbedtls_ssl_context *ssl )
+    switch( mfl )
+    {
+    case MBEDTLS_SSL_MAX_FRAG_LEN_NONE:
+      return MBEDTLS_TLS_EXT_ADV_CONTENT_LEN;
```
Style: `return( MBEDTLS_TLS_EXT_ADV_CONTENT_LEN )`, and likewise below.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -1446,46 +1459,50 @@ run_test    "Max fragment length: enabled, default" \
             -C "found max_fragment_length extension" \
-            -c "16385 bytes written in 2 fragments" \
-            -s "16384 bytes read" \
+            -c "$(expr $MAX_CONTENT_LEN + 1) bytes written in 2 fragments" \
```
Please use shell arithmetic `$((MAX_CONTENT_LEN + 1))`, not `expr`. We don't support pre-POSIX Bourne shells, but we do try to support incomplete Unix environments which may have only `sh` and `grep`. Plus shell arithmetic is faster, especially in environments like Cygwin where fork is very slow.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -608,6 +633,19 @@ make
 msg "test: ssl-opt.sh, MFL-related tests"
 if_build_succeeded tests/ssl-opt.sh -f "Max fragment length"
 
+# ... and with MFL disabled and small output buffer
```
Please make the comment self-contained. You never know when this item is going to be moved, or some other item is going to be inserted before it.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -505,6 +505,30 @@ tests/ssl-opt.sh -f RSA
 tests/compat.sh -t RSA
 
+# Re-run with reduced output message buffer
+msg "build: with ASan, small output buffer"
```
Please put the most important part first and make the message more informative:
```
msg "build: small SSL_OUT_CONTENT_LEN (ASan build)"
msg "test: small SSL_OUT_CONTENT_LEN - ssl-opt.sh MFL-related and large packet tests"
```
(I'd rather drop the `(Asan build)` part of the test description and keep within 79 columns.)

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -505,6 +505,30 @@ tests/ssl-opt.sh -f RSA
+if_build_succeeded tests/ssl-opt.sh -f "Max fragment\|Large packet"
+
+# Re-run MFL tests with both buffers reduced
+msg "build: ASan, small total buffer"
```
Same remark as above regarding leading with the important part. Also this isn't a *total* buffer, it's *both* buffers that are reduced.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -505,6 +505,30 @@ tests/ssl-opt.sh -f RSA
+msg "test: ssl-opt.sh (ASan, MFL-related and large packet tests)"
+if_build_succeeded tests/ssl-opt.sh -f "Max fragment\|Large packet"
+
+# Re-run MFL tests with both buffers reduced
```
I don't think it's that important to test with both buffers reduced. But we should test with a small input buffer and a large output buffer, even if it's a less common configuration, since there's more risk of getting that wrong.

gilles-peskine-arm commented:
`all.sh` and Windows CI: https://jenkins-internal.mbed.com/job/mbedtls-release/504/

projectgus commented:
@sbutcher-arm @gilles-peskine-arm Thanks for the comprehensive comments. Have addressed almost everything, I think. PTAL

> all.sh and Windows CI: https://jenkins-internal.mbed.com/job/mbedtls-release/504/

I don't seem to be able to load this. Do I need to change something?

gilles-peskine-arm commented:
@projectgus Thanks for doing updates so quickly! I'll re-review soon.

Regarding CI results: sorry, that was a note for maintainers only. We can't make our internal CI system visible from the outside for security reasons. You'll have to rely on us to relay results. In this case, it passed, so all is well.

projectgus commented:
Thanks @gilles-peskine-arm :)

Rebased again to resolve conflicts.

sbutcher-arm commented:
And again, I need approval from one of @gilles-peskine-arm / @Patater / @hanno-arm - and then we're (finally!) good to go.

projectgus commented:
@sbutcher-arm *crosses fingers, jumps up and down excitedly* :grin:

sbutcher-arm commented:
Note, on merge #1368 and @pfalcon should be referenced in the ChangeLog for having contributed something very similar, and #1368 should be closed when this is closed.

projectgus commented:
Thanks @sbutcher-arm :smile:

... and with 3 days left before :birthday: :stuck_out_tongue:

pfalcon commented:
Thanks for merging this! Hope a new release will be out soon with this feature ;-).
